### DISCUSSION TITLE
Getting Unknown certificate authority error

### DISCUSSION THREAD

**Discussion by bishtvijay** [gravitational/teleport#42814] [Q&A] [answered]
Hi there, 
I am using community version of teleport,all my services auth and proxy are running in the same ec2 instance when i am trying to configure   storage type: dynamodb i am getting error:
User Message: connection error: desc = &#34;error reading server preface: remote error: tls: unknown certificate authority&#34;] service/connect.go:92
May 09 06:43:54 ip-10-30-254-182 teleport[117990]: 2024-05-09T06:43:54Z ERRO [PROC:1]    Check to see if the config has auth_server pointing to a Teleport Proxy. If it does, use proxy_server instead of auth_server. pid:117990.1 service/connect.go:94

this is my config file
```yaml
version: v3
teleport:
  storage:
    type: dynamodb
    region: 'ap-south-1'
    table_name: 'teleport-one'
    audit_sessions_uri: "s3://teleport/session-recording"
    audit_events_uri:
      - 'dynamodb://teleport-audit-log'
  nodename: ip-10-30-254-182
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  ca_pin: ""
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: bastion.teleport.com
  proxy_listener_mode: multiplex
  public_addr: bastion.teleport.com:3025
ssh_service:
  enabled: "yes"
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr: bastion.teleport.com:443
  https_keypairs: []
  https_keypairs_reload_interval: 0s
  acme:
    enabled: "yes"
    email: vijay.bisht@gmail.com
```

but when i use the same config file but without storage, it is working fine.Can someone help here.

**Comment by leosimony** [comment]
@bishtvijay I am surprised to see another person end with the same issue. Let's hope we will get some help here.

**Comment by leosimony** [comment]
**I've attached my teleport.yaml config file.
I am also running the service in the Amazon EC2 instance.
Teleport version 15.4.6**

```
version: v3
teleport:
  nodename: teleport-dn 
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
      extra_fields: [level, timestamp, component, caller] 
  ca_pin: ""
  diag_addr: ""
  storage:
   type: dynamodb
   region: ap-south-2
   table_name: teleport-backend
   audit_events_uri:  ['dynamodb://teleport-audit', 'file:///var/lib/teleport/audit/events', 'stdout://']
   audit_sessions_uri: "s3://teleport-sesssion/"
   retention_period: 365d
   billing_mode: "pay_per_request"
   continuous_backups: true
   proxy_server: teleport.devops.com:443

auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: teleport.devops.com
  proxy_listener_mode: multiplex
  proxy_protocol: off
  public_addr: teleport.devops.com:3025

ssh_service:
  enabled: "yes"
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s

proxy_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3023
  tunnel_listen_addr: 0.0.0.0:3024
  web_listen_addr: 0.0.0.0:443
  public_addr: teleport.devops.com:443
  https_keypairs: []
  https_keypairs_reload_interval: 0s
  acme:
    enabled: "yes"
    email: leosimon@devops.com
```

**After I made the changes to the dynamodb storage backend, I ended up with the below errors and if I remove the storage backend configurations, everything works fine.**

```
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: 2024-07-04T12:18:33Z INFO [PROC:1]    Connecting to the cluster with TLS client certificate. pid:1761.1 cluster:teleport.devops.com service/connect.go:238
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: 2024-07-04T12:18:33Z WARN [MXTLS:1]   Handshake failed. dst_addr:127.0.0.1:3025 error:[tls: failed to verify certificate: x509: certificate signed by unknown authority] src_addr:127.0.0.1:43282 multiplexer/tls.go:152
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: 2024-07-04T12:18:33Z ERRO [PROC:1]    Failed to establish connection to cluster. pid:1761.1 identity:Instance error:[
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: ERROR REPORT:
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: Original Error: trace.aggregate connection error: desc = &#34;error reading server preface: remote error: tls: unknown certificate authority&#34;
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: Stack Trace:
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/connect.go:1320 github.com/gravitational/teleport/lib/service.(*TeleportProcess).newClientDirect
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/connect.go:1168 github.com/gravitational/teleport/lib/service.(*TeleportProcess).newClient.func1
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/connect.go:1237 github.com/gravitational/teleport/lib/service.(*TeleportProcess).newClient
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/connect.go:1126 github.com/gravitational/teleport/lib/service.(*TeleportProcess).getConnector
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/connect.go:239 github.com/gravitational/teleport/lib/service.(*TeleportProcess).connect
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/connect.go:188 github.com/gravitational/teleport/lib/service.(*TeleportProcess).connectToAuthService
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/connect.go:81 github.com/gravitational/teleport/lib/service.(*TeleportProcess).reconnectToAuthServiceJul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/service.go:3027 github.com/gravitational/teleport/lib/service.(*TeleportProcess).RegisterWithAuthServer.func1
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/supervisor.go:587 github.com/gravitational/teleport/lib/service.(*LocalService).Serve
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         github.com/gravitational/teleport/lib/service/supervisor.go:312 github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func1
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]:         runtime/asm_amd64.s:1695 runtime.goexit
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: User Message: connection error: desc = &#34;error reading server preface: remote error: tls: unknown certificate authority&#34;] service/connect.go:97
Jul 04 12:18:33 ip-172-31-4-203 teleport[1761]: 2024-07-04T12:18:33Z ERRO [PROC:1]    Check to see if the config has auth_server pointing to a Teleport Proxy. If it does, use proxy_server instead of auth_server. pid:1761.1 service/connect.go:99
```

**Reply by webvictim** [reply]
@leosimony Your issue appears to be that you are trying to run the `auth_service` and `proxy_service` locally, but you are _also_ specifying `proxy_server: teleport.devops.com:443` in your config to tell Teleport to make a connection to a remote server.

You can only do one or the other. If you remove `proxy_server: teleport.devops.com:443` from your config and restart, it should work fine.

If this doesn't work, please try the other solution I wrote below.

**Reply by leosimony** [reply]
Exactly. I did not add proxy_server earlier. It's a long story.
I was trying teleport setup with docker-compose env with the nginx proxy manager. Since the error message appeared after changing the storage type, I searched in forums and tried multiple things. Finally, I ended up setting up a stand-alone system at AWS EC2 and wanted to check how it's working. Looks like I retained the old configs. 
Thanks a ton for your hep, much appreciated.

**Comment by webvictim** [comment, marked as answer]
This is possibly an issue due to existing local state being kept in `/var/lib/teleport/proc` and interfering with the cluster.

If you're looking to set up a new cluster with a DynamoDB backend, try this:
- stop the Teleport process
- delete the contents of the Teleport data directory: `rm -rf /var/lib/teleport`
- start the Teleport process again

Note that by doing this, you will lose all your configured users, roles etc - but migrating data from a local backend to DynamoDB is non-trivial, so starting fresh is best.

**Reply by leosimony** [reply]
This seems to be the key to all the issues I have faced since changing the storage type. Tried setting up teleport with SQlite and changing to dynamodb is not compatible it looks like. We need to do a fresh to dynamodb itself. As you suggested, I have deleted everything in the data directory and a simple service restart fixed everything. I hope its mentioned in the official documentation and I missed it. If not, its better if it gets updated with this step. Fortunately, I am trying to setup teleport for our environment and in the starting stage only, no worries about data loss etc.
Once again, much appreciated for your suggestion. Literally you made my day.

**Reply by webvictim** [reply]
This is actually related to one of our oldest issues: https://github.com/gravitational/teleport/issues/2838

It's hard to reliably fix the issue without manual intervention while always preserving the state of existing clusters where needed, so I think we've erred on the side of just leaving people to either discover the fix for themselves or advising them when they ask.

My understanding is that this issue will affect any cluster migrating from a local sqlite backend to another backend, so I will take a look at the backend docs to see if there's a logical place to add a warning.

**Comment by leosimony** [comment]
@bishtvijay Hope the steps provided will help resolve your issue too.
